### ISSUE TITLE
About STDP code implementation

### ISSUE THREAD

**Issue by 932179209** [BrainCog-X/Brain-Cog#26]
Hi!
I'm reading STDP code in braincog.base.learningrule.
But it seems like the way to implement the STDP is not the same as the formula (9) in the Braincog thesis.

`class STDP(nn.Module):
    """
    STDP learning rule
    """

    def __init__(self, node, connection, decay=0.99):
        """
        :param node:node神经元类型实例如IFNode LIFNode
        :param connection:连接 类的实例 里面只能有一个操作
        """
        super().__init__()

        self.node = node
        self.connection = connection
        self.trace = None
        self.decay = decay

    def forward(self, x):
        """
        计算前向传播过程
        :return:s是脉冲 dw更新量
        """
        x = x.clone().detach()
        i = self.connection(x)
        with torch.no_grad():
            s = self.node(i)

            i.data += s - i.data
            trace = self.cal_trace(x)
            x.data += trace - x.data

        dw = torch.autograd.grad(outputs=i, inputs=self.connection.weight, grad_outputs=i)

        return s, dw

    def cal_trace(self, x):
        """
        计算trace
        """
        if self.trace is None:
            self.trace = Parameter(x.clone().detach(), requires_grad=False)
        else:
            self.trace *= self.decay
            self.trace += x
        return self.trace.detach()

    def reset(self):
        """
        重置
        """
        self.trace = None`

I have some doubts, specifically, 
1. there is no Δt in the code, how to calculate W(Δt)？
2. the 'trace' seems like unused in the forward pass, what does 'trace' mean in the code?
3. the STDP is a unsupervised learning rule, but the dw is calculate by torch.autograd.grad, how it works and why use that?

Thanks for your time!

**Comment by Ikarosy** [comment]
I share the same Q.1 and Q.2 as commentator 932179209.  For Q.3, I think the implementation code does not use any supervised gradient from the loss function w.r.t. LABEL, which matches the philosophy of unsupervised learning rules.  But I don't understand why 'trace' is not involved in calculating 'dw.'  As far as I know, STDP usually requires the 'trace' to update weights.  Could the code maintainer upload more explanations or papers related to this implementation code?  It will be of great help!  

I am looking forward to your reply.

**Comment by Ikarosy** [comment]
Additionally, I find the code implementation of STDP here is also different from [SpikingJelly](https://github.com/fangwei123456/spikingjelly/blob/master/spikingjelly/activation_based/learning.py).  Does it mean there exist more different STDP variants?  Will BrainCog support all these STDP variants?

Thanks for your great effort. Hope to see more exciting improvements in the future.

**Comment by yizeng-ai** [comment]
> Additionally, I find the code implementation of STDP here is also different from [SpikingJelly](https://github.com/fangwei123456/spikingjelly/blob/master/spikingjelly/activation_based/learning.py). Does it mean there exist more different STDP variants? Will BrainCog support all these STDP variants?
> 
> Thanks for your great effort. Hope to see more exciting improvements in the future.

There are several STDP variations, both computationally and biologically, and we are in the process of supporting these different variations, not only to support modeling the brains, while more importantly to explore how these different variations contribute to different aspects of Brain-inspired AI models.

**Comment by sunbaby01** [comment]
> Hi! I'm reading STDP code in braincog.base.learningrule. But it seems like the way to implement the STDP is not the same as the formula (9) in the Braincog thesis.
> 
> `class STDP(nn.Module): """ STDP learning rule """
> 
> ```
> def __init__(self, node, connection, decay=0.99):
>     """
>     :param node:node神经元类型实例如IFNode LIFNode
>     :param connection:连接 类的实例 里面只能有一个操作
>     """
>     super().__init__()
> 
>     self.node = node
>     self.connection = connection
>     self.trace = None
>     self.decay = decay
> 
> def forward(self, x):
>     """
>     计算前向传播过程
>     :return:s是脉冲 dw更新量
>     """
>     x = x.clone().detach()
>     i = self.connection(x)
>     with torch.no_grad():
>         s = self.node(i)
> 
>         i.data += s - i.data
>         trace = self.cal_trace(x)
>         x.data += trace - x.data
> 
>     dw = torch.autograd.grad(outputs=i, inputs=self.connection.weight, grad_outputs=i)
> 
>     return s, dw
> 
> def cal_trace(self, x):
>     """
>     计算trace
>     """
>     if self.trace is None:
>         self.trace = Parameter(x.clone().detach(), requires_grad=False)
>     else:
>         self.trace *= self.decay
>         self.trace += x
>     return self.trace.detach()
> 
> def reset(self):
>     """
>     重置
>     """
>     self.trace = None`
> ```
> 
> I have some doubts, specifically,
> 
> 1. there is no Δt in the code, how to calculate W(Δt)？
> 2. the 'trace' seems like unused in the forward pass, what does 'trace' mean in the code?
> 3. the STDP is a unsupervised learning rule, but the dw is calculate by torch.autograd.grad, how it works and why use that?
> 
> Thanks for your time!

Thanks for your question. 
The formula you said is the conventional STDP. There are some variations. For the computational efficiency, we use Eligibility Traces. As shown in follow:

$$
	\begin{split}
		&\Delta w_{j}=x_{trace}^{(t)}\quad if  \quad x_{postspike}=1\\
		&\begin{split}
			x_{trace}^{(t)}=\lambda_{+}x_{trace}^{(t-1)}+x_{prespike}
		\end{split}
	\end{split}
$$

Several works used this variation. And this is unilateral STDP. A bilateral version is in class fullstdp.

In implementation, we use computational graph supported by pytorch.  we replace the x of graph with trace in   `x.data += trace - x.data`, and the dy of graph with s in `i.data += s - i.data`
In equation dw=x\*dy->dw=trace\*s. 

More detail about computational graph in pytorch.org.

**Comment by Ikarosy** [comment]
> > Additionally, I find the code implementation of STDP here is also different from [SpikingJelly](https://github.com/fangwei123456/spikingjelly/blob/master/spikingjelly/activation_based/learning.py). Does it mean there exist more different STDP variants? Will BrainCog support all these STDP variants?
> > Thanks for your great effort. Hope to see more exciting improvements in the future.
> 
> There are several STDP variations, both computationally and biologically, and we are in the process of supporting these different variations, not only to support modeling the brains, while more importantly to explore how these different variations contribute to different aspects of Brain-inspired AI models.

Thanks for your reply, can't wait to see your future updates.
